File: coding-interviewes/coding_interview_150/scala/BestTimeToBuyAndSellStockII.scala

/* Best Time II — sum positive deltas. O(n) */
object BestTimeToBuyAndSellStockII {
  def maxProfit(prices:Array[Int]):Int = {
    var ans=0
    for (i <- 1 until prices.length) if (prices(i)>prices(i-1)) ans += prices(i)-prices(i-1)
    ans
  }
  def main(args:Array[String]):Unit = {
    println(maxProfit(Array(7,1,5,3,6,4)))
    println(maxProfit(Array(1,2,3,4,5)))
  }
}

File: coding-interviewes/coding_interview_150/java/BestTimeToBuyAndSellStockII.java

/* Best Time II — sum increases. O(n) time, O(1) space */
public class BestTimeToBuyAndSellStockII {
    public static int maxProfit(int[] prices){
        int ans=0; for (int i=1;i<prices.length;i++) if (prices[i]>prices[i-1]) ans += prices[i]-prices[i-1];
        return ans;
    }
    public static void main(String[] args){
        System.out.println(maxProfit(new int[]{7,1,5,3,6,4}));
        System.out.println(maxProfit(new int[]{1,2,3,4,5}));
    }
}
